sparse/core.py

```diff
@@ -603,55 +613,238 @@ def elemwise_binary(self, func, other, *args, **kwargs):
+            coords_list.append(other_unmatched_coords)
+            data_list.append(func(0, other_broadcast_data,
+                                  *args, **kwargs))
+            coords_list.append(other_broadcast_coords)
 
         # Concatenate matches and mismatches
-        data = np.concatenate([func(self_data[matched_self],
-                                    other_data[matched_other],
-                                    *args, **kwargs),
-                               func(self_data[unmatched_self], 0,
-                                    *args, **kwargs),
-                               func(0, other_data[unmatched_other],
-                                    *args, **kwargs)])
-        coords = np.concatenate([self_coords[matched_self],
-                                 self_coords[unmatched_self],
-                                 other_coords[unmatched_other]])
+        data = np.concatenate(data_list) if len(data_list) else np.empty((0,), dtype=self.dtype)
+        coords = np.concatenate(coords_list, axis=1) if len(coords_list) else np.empty((0, len(result_shape)),
+                                                                                       dtype=self.coords.dtype)
 
         nonzero = data != 0
         data = data[nonzero]
-        coords = coords[nonzero]
+        coords = coords[:, nonzero]
+
+        return COO(coords, data, shape=result_shape, has_duplicates=False)
+
+    @staticmethod
+    def _get_broadcast_shape(shape1, shape2, is_result=False):
+        """
+        Get the overall broadcasted shape from two different shapes, without actually constructing any arrays.
+        """
+        # https://stackoverflow.com/a/47244284/774273
+        if not all((l1 == l2) or (l1 == 1) or ((l2 == 1) and not is_result) for l1, l2 in zip(shape1[::-1],
+                                                                                             shape2[::-1])):
+            raise ValueError('operands could not be broadcast together with shapes %s, %s' % (shape1, shape2))
+
+        broadcast_shape = tuple(max(l1, l2) for l1, l2 in zip_longest(shape1[::-1], shape2[::-1], fillvalue=1))[::-1]
+
+        return broadcast_shape
+
+    @staticmethod
+    def _get_broadcast_parameters(shape, broadcast_shape):
+        """
+        Get broadcast parameters, i.e. tuple with None/True/False. None if extra dimension is added, True if the
+        dimension doesn't need broadcasting, and False if it needs broadcasting.
+        """
+        params = [None if l1 is None else l1 == l2 for l1, l2 in zip_longest(shape[::-1], broadcast_shape[::-1],
+                                                                             fillvalue=None)][::-1]
+
+        return params
+
+    @staticmethod
+    def _get_reduced_coords(coords, params):
+        reduced_params = [bool(param) for param in params]
+
+        return coords[reduced_params]
+
+    @staticmethod
+    def _get_expanded_coords_data(coords, data, params, broadcast_shape):
```

Comment:
The purpose of this function isn't immediately clear to me.  Can I ask you for a docstring to aid review?  This would generally be appreciated for similar methods as well.

Reply:
I'll add doctstrings for all methods.